test/python/circuit/test_parameters.py

```diff
@@ -170,3 +170,73 @@ def test_parameter_name_conflicts_raises(self):
         qc.u1(theta1, 0)
 
         self.assertRaises(QiskitError, qc.u1, theta2, 0)
+
+    def test_bind_ryrz_vector(self):
+        """Test binding a list of floats to a ParamterVector"""
+        qc = QuantumCircuit(4)
+        depth = 4
+        theta = ParameterVector('θ', length=len(qc.qubits) * depth * 2)
+        theta_iter = iter(theta)
+        for _ in range(depth):
+            for q in qc.qubits:
+                qc.ry(next(theta_iter), q)
+                qc.rz(next(theta_iter), q)
+            for i, q in enumerate(qc.qubits[:-1]):
+                qc.cx(qc.qubits[i], qc.qubits[i+1])
+            qc.barrier()
+        theta_vals = numpy.linspace(0, 1, len(theta)) * numpy.pi
+        self.assertEqual(set(qc.parameters), set(theta.params))
+        bqc = qc.bind_parameters({theta: theta_vals})
+        for gate_tuple in bqc.data:
+            if hasattr(gate_tuple[0], 'params') and gate_tuple[0].params:
+                self.assertIn(gate_tuple[0].params[0], theta_vals)
+
+    def test_compile_vector(self):
+        """Test compiling a circuit with an unbound ParamterVector"""
+        qc = QuantumCircuit(4)
+        depth = 4
+        theta = ParameterVector('θ', length=len(qc.qubits)*depth*2)
+        theta_iter = iter(theta)
+        for _ in range(depth):
+            for q in qc.qubits:
+                qc.ry(next(theta_iter), q)
+                qc.rz(next(theta_iter), q)
+            for i, q in enumerate(qc.qubits[:-1]):
+                qc.cx(qc.qubits[i], qc.qubits[i+1])
+            qc.barrier()
+        backend = BasicAer.get_backend('qasm_simulator')
+        qc_aer = transpile(qc, backend)
+        for param in theta:
+            self.assertIn(param, qc_aer.parameters)
+
+    def test_instruction_ryrz_vector(self):
+        """Test constructing a circuit from instructions with remapped ParamterVectors"""
+        qubits = 5
+        depth = 4
+        ryrz = QuantumCircuit(qubits, name='ryrz')
+        theta = ParameterVector('θ0', length=len(ryrz.qubits) * 2)
+        theta_iter = iter(theta)
```

Comment:
I think you can make the ParameterVector object itself iterable (just define `__iter__` and `__next__`). That way you don't have to explicitly convert to iterable here, and you could do things like `for single_theta in theta: ...`. This seems more aligned with the goal of making this class easy to work with and not having to rely on Python loops, iterations, etc.

Reply:
The only reason I have the `iter` call here is because I call `next` twice inside the loop, and for some reason there's no clean way that I know of to iterate in a for loop in buckets of n items (e.g. for each step give me four items). The iteration you want works here too, because when you call `for theta_n in theta` it will call `iter(theta)`, get the list iterator, and then each theta will be the `Parameter` objects in the list. I originally had `next()` implemented and it was just calling `next()` on a list iterator which was acquired and saved inside `ParameterVector.__iter__()` but realized we don't need it and it's a little messy. Does that make sense?

Reply:
I just pushed [61b1067](https://github.com/Qiskit/qiskit-terra/pull/2379/commits/61b1067a5833d64243ab40bf666ca3e0637d0663) to your branch, which I think does what I'm saying here. Basically now you can call `iter(theta_vector)` -- it behaves like a normal iterator. And the explicit conversion to `iter` is not needed anymore.

Reply:
The problem with making `ParameterVector` the an iterator is that iterators are inherently single-use.

```
>>> v = ParameterVector('v', 3)
>>> next(v)
Parameter(v[0])
>>> next(v)
Parameter(v[1])
>>> next(v)
Parameter(v[2])
>>> next(v)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/kevin.krsulichibm.com/q/qiskit-terra/qiskit/circuit/parametervector.py", line 49, in __next__
    raise StopIteration
StopIteration
```
Now there's no way to go back and reset the iterator over `v`. For example, `List`s are not interators, but you can generate an interator from a `List`.

```
>>> b = [1,2,3]
>>> next(b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'list' object is not an iterator
>>> i = iter(b)
>>> next(i)
1
>>> next(i)
2
>>> i2 = iter(b)
>>> next(i2)
1
>>> next(i)
3
```

Prior to https://github.com/Qiskit/qiskit-terra/pull/2379/commits/61b1067a5833d64243ab40bf666ca3e0637d0663 , it was still possible to write `for param in vector: ...` since `ParameterVector.__iter__` would return an iterator over `self._params`, which would have it's own `__iter__` and `__next__`.

The reason the RyRz test explicitly calls `iter(ParameterVector)` is because its implemented by pulling off parameters one by one in sequence, rather than by drawing groups of parameters at once.

e.g. If I have a a group of three gates I'd like to repeat five times, I could write either:
```
v = ParameterVector('v', 15)
v_iter = iter(v)
for group_index in range(5):
  qc.rx(next(v_iter), 0)
  qc.ry(next(v_iter), 0)
  qc.rz(next(v_iter), 0)
```
or which, to me is the python equivalent of something like:
```
v = ParameterVector('v', 15)
i = 0
while i < len(v):
  qc.rx(v[i++], 0)
  qc.ry(v[i++], 0)
  qc.rz(v[i++], 0)
```

@dongreenberg is using `iter` as a means of keeping track of position within the `ParameterVetor`, not because it didn't natively support iteration.

Reply:
Depends on whether you want this object to behave purely like a list, or to also be able to use it as an iterator. If you think that being able to do the following is useful, then `__next__` does that.
```python
In [2]: v = ParameterVector('v', 3)

In [3]: next(v)
Out[3]: Parameter(v[0])
```
Note that this does not preclude the previous behavior, i.e. you can still do:
```python
In [4]: i = iter(v)

In [5]: next(i)
Out[5]: Parameter(v[0])

In [6]: next(i)
Out[6]: Parameter(v[1])

In [7]: i2 = iter(v)

In [8]: next(i2)
Out[8]: Parameter(v[0])

In [9]: next(i)
Out[9]: Parameter(v[2])
```
But I'm fine reverting the commit, it was mostly to simplify those tests.

Reply:
Oh oh oh I didn't catch that. I like that, but leave it up to kdk because I'm a shorthand liberal.

Reply:
I like the conciseness, but to me, it pushes `ParameterVector` too far from convention. It's a list-like thing, and you can use it in most of the standard like-like ways. (e.g. "If you want it as an interator, use `iter`.") Also, that it can only be single use as an iterator will only ever confuse someone who tries to use it that way.

(As an aside, it looks like an iterator is [expected](https://docs.python.org/3/library/stdtypes.html#iterator.__iter__) to return `self` on it's `__iter__` method, though I can't off the top of my head see what this would break.)

Reply:
Ok I'm fine with having this object as just a list. I just reverted the commit.